model.py

```
@@ -177,52 +254,76 @@ def train_or_update_variational_model(
+
+            # Invert the normalization on our outputs
+            mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
+            untransformed_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
+
+            var_scaler = sp.label_scaler.named_transformers_['var_minmax_scaler']
+            unscaled_var = var_output * var_scaler.scale_ + var_scaler.min_
+            untransformed_var = SoftplusLayer()(unscaled_var)
+
+            # Output mean, tuples.
+            outputs = keras.layers.concatenate([untransformed_mean, untransformed_var])
+            model = keras.Model(inputs=inputs, outputs=outputs)
+
+            optimizer = keras.optimizers.Adam(learning_rate=lr)
+            double_sided_kl_tf = tf.constant(double_sided_kl)
+            num_samples_tf = tf.constant(kl_num_samples_from_pred_dist)
+            model.compile( 
+                optimizer=optimizer, 
+                loss=KLCustomLoss(double_sided_kl_tf, num_samples_tf))
+            model.summary()
+        else:
+            model = keras.models.load_model(
+                model_file,
+                custom_objects={"KLCustomLoss": KLCustomLoss})
+        model.save(model_file)
+        dump(sp.feature_scaler, f"{model_file.strip('.keras')}.pkl")
+        packaged_model = TFModel(model_file, f"{model_file.strip('.keras')}.pkl")
+        return packaged_model
+    
+    if sp.val: 
+        # No cross validation
+        assert not n_cv_folds, "Cross validation not possible if manually specifying a validation dataset."
+        model = build_model()
+        history = model.fit(sp.train.X, sp.train.Y, verbose=0, validation_data=sp.val.as_tuple(), shuffle=True,
+                            epochs=epochs, batch_size=batch_size, callbacks=callbacks_list)
+        return history, model
+    else:
+        # Yes cross validation
+        assert n_cv_folds, "If no validation dataset is specified, number of cross validation folds must be set."
```

Prefer to raise an exception-- asserts are only guaranteed to be evaluated in debug mode.